packages/server/lib/pki.ts

```diff
@@ -0,0 +1,234 @@
+              new networkPki.PfxCertificate(pfxRaw, passphrase),
+            )
+
+            const subject = extractSubjectFromPfx(pfxParsed)
+
+            urlPkiCertificates.addSubject(subject)
+            debug(
+              `loaded client PFX certificate: ${subject} for url: ${urlPkiCertificates.url}`,
+            )
+          }
+        })
+
+        clientPkiCertificateStore.addPkiCertificatesForUrl(urlPkiCertificates)
+        index++
+      })
+
+      debug(
+        `loaded client PKI certificates for ${clientPkiCertificateStore.getCertCount()} URL(s)`,
+      )
+    }
+  } catch (e) {
+    debug(
+      `Failed to load client PKI certificate for clientPkiCertificates[${index}]: ${e.message} ${e.stack}`,
+    )
+
+    throw new Error(
+      `Failed to load client PKI certificates for clientPkiCertificates[${index}]: ${e.message}.  For more debug details run Cypress with DEBUG=cypress:server:pki*`,
+    )
+  }
+}
+
+function createAbsoluteFilepath (root: string, filepath: string): string {
+  if (!filepath) {
+    throw new Error('filepath is not defined')
+  }
+
+  if (filepath.startsWith('/')) {
+    return filepath
+  }
+
+  return path.isAbsolute(filepath) ? filepath : path.join(root, filepath)
+}
+
+function loadBinaryFromFile (projectRoot: string, filepath: string): Buffer {
+  let retrievePath: string = createAbsoluteFilepath(projectRoot, filepath)
+
+  debug(`loadCertificateFile: ${retrievePath}`)
+
+  return fs.readFileSync(retrievePath)
+}
+
+function loadTextFromFile (projectRoot: string, filepath: string): string {
+  let retrievePath: string = createAbsoluteFilepath(projectRoot, filepath)
+
+  debug(`loadPassphraseFile: ${retrievePath}`)
+
+  return fs.readFileSync(retrievePath, 'utf8').toString()
+}
+
+/**
+ * Extract subject from supplied pem instance
+ */
+function extractSubjectFromPem (pem): string {
+  try {
+    return pem.subject.attributes
+    .map((attr) => [attr.shortName, attr.value].join('='))
+    .join(', ')
+  } catch (e) {
+    throw new Error(`Unable to extract subject from PEM file: ${e.message}`)
+  }
+}
+
+/**
+ * Load PFX data from the supplied Buffer and passphrase
+ */
+function loadPfx (pfx: Buffer, passphrase: string | undefined) {
+  try {
+    const certDer = util.decode64(pfx.toString('base64'))
+    const certAsn1 = asn1.fromDer(certDer)
+
+    return pkcs12.pkcs12FromAsn1(certAsn1, passphrase)
+  } catch (e) {
+    debug(`loadPfx fail: ${e.message} ${e.stackTrace}`)
+    throw new Error(`Unable to load PFX file: ${e.message}`)
+  }
+}
+
+/**
+ * Extract subject from supplied pfx instance
+ */
+function extractSubjectFromPfx (pfx) {
+  try {
+    const certs = pfx.getBags({ bagType: pki.oids.certBag })[pki.oids.certBag].map((item) => item.cert)
+
+    return certs[0].subject.attributes.map((attr) => [attr.shortName, attr.value].join('=')).join(', ')
+  } catch (e) {
+    throw new Error(`Unable to extract subject from PFX file: ${e.message}`)
+  }
+}
```

Comment:
this logic, and other logic in this file that uses `node-forge`, seems like it belongs in `@packages/network`, not here. would you agree?